common.py

```diff
@@ -10,6 +10,69 @@
 from typing import cast
 
 
+# This is used as default value for some arguments, to recognize that the
+# caller didn't specify the argument. This is useful, when we want to
+# explicitly distinguish between having an argument unset or set to any value.
+# The caller would never pass this value, but the implementation would check
+# whether the argument is still left at the default.
+#
+# See also, dataclasses.MISSING and dataclasses._MISSING_TYPE
+class _MISSING_TYPE:
+    pass
+
+
+MISSING = _MISSING_TYPE()
+
+
+def bool_to_str(val: bool, *, format: str = "yes") -> str:
+    if format == "yes":
+        return "yes" if val else "no"
+    raise ValueError(f'Invalid format "{format}"')
+
+
+T1 = TypeVar("T1")
+T2 = TypeVar("T2")
+
+
+def str_to_bool(
+    val: None | str | bool,
+    on_error: T1 | _MISSING_TYPE = MISSING,
+    *,
+    on_default: T2 | _MISSING_TYPE = MISSING,
+) -> bool | T1 | T2:
```

Comment:
I don't think I understand what is being added by having Missing_Type.

In my mind str_to_boot() should return a bool, and raise an exception if for some reason it is called with a str that cannot be intuitively converted to a bool.

Can you elaborate why this is necessary?

Reply:
> In my mind str_to_boot() should return a bool, and raise an exception if for some reason it is called with a str that cannot be intuitively converted to a bool.

That's what it does. `common.str_to_bool("bogus")` raises an exception.

However, if you do `common.str_to_bool("bogus", on_error=something)` (or simply `common.str_to_bool("bogus", something)`, then on error `something` is returned.

the `_MISSING_TYPE` is there to get the typing information right.

- If you call `common.str_to_bool(str)`, then mypy knows that this (only) returns a `bool`. That means, the caller doesn't need to further check either.

- If you call `common.str_to_bool(str, something)`, then mypy knows that this returns `bool | type(something)`, consequently, the caller will be forced to deal with that. For example,

For example,
```python
    var: bool
    var = common.str_to_bool(str, None)

    ternary_val: Optional[bool]
    ternary_val = common.str_to_bool(str, None)
```
both works, as you would would expect. And mypy also will agree with this.

For example
```python
   def func(val: bool)
       pass

   func(common.str_to_bool(str, None))
```
will be rejected by the linter, because you didn't handle the `None`.